doc/dev/osd_internals/erasure_coding/enhancements.rst

```diff
@@ -0,0 +1,1517 @@
+end up with a lot of CRC metadata) and updating these CRCs when data is
+overwritten (the update would need to perform a read-modify-write at the same
+granularity as the CRC). These finer granularity CRCs can then easily be
+combined to produce a CRC for the whole shard or even the whole erasure coded
+object.
+
+This proposal suggests going in the opposite direction - EC overwrite pools
+have survived without CRCs and relied on Bluestore up until now, so why is
+this feature needed? The current code doesn’t check CRCs if overwrite is
+enabled, but sadly still calculates and updates a CRC in the hinfo xattr, even
+if performing overwrites which mean that the calculated value will be
+garbage. This means we pay all the overheads of calculating the CRC and get no
+benefits.
+
+The code can easily be fixed so that CRCs are calculated and maintained when
+objects are written sequentially, but as soon as the first overwrite to an
+object occurs the hinfo xattr will be discarded and CRCs will no longer be
+calculated or checked. This will improve performance when objects are
+overwritten, and will improve data integrity in cases where they are not.
+
+While the thought is to abandon EC storing CRCs in objects being overwritten,
+there is an improvement that can be made to deep scrub. Currently deep scrub
+of an EC with overwrite pool just checks that every shard can read the object,
+there is no checking to verify that the copies on the shards are consistent. A
+full consistency check would require large data transfers between the shards
+so that the coding parities could be recalculated and compared with the stored
+versions, in most cases this would be unacceptably slow. However for many
+erasure codes (including the default ones used by Ceph) if the contents of a
+chunk are XOR’d together to produce a longitudinal summary value, then an
+encoding of the longitudinal summary values of each data shard should produce
+the same longitudinal summary values as are stored by the coding parity
+shards. This comparison is less expensive than the CRC checks performed by
+replication pools. There is a risk that by XORing the contents of a chunk
+together that a set of corruptions cancel each other out, but this level of
+check is better than no check and will be very successful at detecting a
+dropped write which will be the most common type of corruption.
+
+Metadata changes
+----------------
+
+What metadata do we need to consider?
+
+1. object_info_t. Every Ceph object has some metadata stored in the
+   object_info_t data structure. Some of these fields (e.g. object length) are
+   not updated frequently and we can simply avoid performing partial writes
+   optimizations when these fields need updating. The more problematic fields
+   are the version numbers and the last modification time which are updated on
+   every write. Version numbers of objects are compared to version numbers in
```

Comment:
We may need to extend `object_info_t` with a bitmap to implement sparse reads (`CEPH_OSD_OP_SPARSE_READ`).  The bitmap would need to be updated on each write operation that changes the allocation status of the block, with the assumption that ops like `CEPH_OSD_OP_TRUNCATE` and `CEPH_OSD_OP_ZERO` deallocate.  Not sure if this can be made to work with "Direct reads" as proposed in this doc or sparse reads would just be excluded and would always need to be sent to the primary OSD.

Context: Precise allocation information down to 4096 or even 512-byte blocks is needed for built-in LUKS encryption feature in RBD and filesystem-level encryption (fscrypt) support in CephFS to avoid "decrypting" a buffer with synthesized zeroes and returning the resulting gibberish to the user.  Currently, `CEPH_OSD_OP_SPARSE_READ` works only on replicated pools -- there is no support in the OSD per se, instead it's based on responses from the object store (`ObjectStore::fiemap()`).  On EC pools, since the object is sharded and the object store on the primary OSD can't respond with anything useful, sparse reads are degraded into regular reads (i.e. those that synthesize zeroes):

https://github.com/ceph/ceph/blob/8017ef589d2009df0f1e984ac8cc5b41a0ed9189/src/osd/PrimaryLogPG.cc#L5918-L5932

This means that neither built-in LUKS encryption in RBD nor fscrypt in CephFS can be fully supported on EC pools.  I'd love us to change that the patient is opened up for surgery, so to speak.

cc @athanatos @rzarzynski @markhpc

Reply:
Technically, an OSD is allowed to fall back to a regular read because `ObjectStore::fiemap()` and `CEPH_OSD_OP_SPARSE_READ` were initially added as an optimization.  In fact, an object store implementation that always lies by responding with "this extent is fully allocated" would be deemed compliant:
https://github.com/ceph/ceph/blob/8017ef589d2009df0f1e984ac8cc5b41a0ed9189/src/os/ObjectStore.h#L492-L499

However, with Bluestore on replicated pools this never happens, so over the years we have grown users that use `CEPH_OSD_OP_SPARSE_READ` for more than just reducing the amount of network traffic needed to transmit synthesized zeroes.  Promoting `CEPH_OSD_OP_SPARSE_READ` from something that is best-effort to something that is guaranteed is highly desired and moving it up into the OSD so that it's independent of the object store and works the same way on both replicated and EC pools seems to be the only way to get there.

Reply:
Yep.  I'm writing up a tracker feature for this.  I'd like to go ahead and merge this PR as is and follow up with a plan to deal with the fiemap/sparse read issue in a future PR.

Reply:
https://tracker.ceph.com/issues/67507